File: test/exception/FExceptionAggregate.test.ts
import { FException, FExceptionAggregate } from "../../src/index";

import { assert } from "chai";

describe("FExceptionAggregate test", function () {
	it("should be instantable without inner errors", function () {
		const aggrError = new FExceptionAggregate([]);
		assert.equal(aggrError.message, FExceptionAggregate.name);
	});

	it("should concatenate messages from inner errors", function () {
		let err1: FException;
		let err2: FException;
		let err3: FException;

		try { throw new FException("Err1"); } catch (e) { err1 = e as FException; }
		try { throw new FException("Err2"); } catch (e) { err2 = e as FException; }
		try { throw new FException("Err3"); } catch (e) { err3 = e as FException; }

		const aggrError = new FExceptionAggregate([err1, err2, err3]);

		assert.equal(aggrError.message, "Err1\nErr2\nErr3");
		assert.equal(aggrError.toString(), `${err1.toString()}\n${err2.toString()}\n${err3.toString()}`);
	});
});

File: test/exception/f_exception_aggregate_test.dart
// Copyright 2023, the Freemework.ORG project authors. Please see the AUTHORS
// file for details. All rights reserved. Use of this source code is governed
// by a BSD-style license that can be found in the LICENSE file.

import "dart:convert" show LineSplitter;

import "package:test/test.dart";

import "package:freemework/freemework.dart";

void main() {
  group("FExceptionAggregate tests", () {
    setUp(() {});

    tearDown(() {});

    test("should be instantable without inner errors", () {
      FExceptionAggregate([]); // no expection expected
    });

    test("should be instantable without inner errors", () {
      FExceptionAggregate aggrError;
      try {
        throw FExceptionAggregate([], "my message");
      } catch (e) {
        aggrError = e as FExceptionAggregate;
      }
      expect(aggrError.message, equals("my message"));
    });

    test("should concatenate messages from inner errors", () {
      FException err1;
      FException err2;
      FException err3;

      try {
        throw FException("Err1");
      } catch (e) {
        err1 = e as FException;
      }
      try {
        throw FException("Err2");
      } catch (e) {
        err2 = e as FException;
      }
      try {
        throw FException("Err3");
      } catch (e) {
        err3 = e as FException;
      }

      FExceptionAggregate aggrError;
      try {
        throw FExceptionAggregate([err1, err2, err3]);
      } catch (e) {
        aggrError = e as FExceptionAggregate;
      }

      expect(aggrError.message, equals("One or more errors occurred."));
      expect(
          LineSplitter.split(aggrError.toString()).first,
          equals(
            "FExceptionAggregate: One or more errors occurred. ---> FException: Err1",
          ));
    });
  });
}
